bindings/python/linkgrammar.py

```diff
@@ -81,210 +38,306 @@ def __init__(self, verbosity = 0,
-
-        # XXX FIXME: doing the below, if not needed, is a bit of a cpu-waster!
-        #
-        # display_walls=0, screen_width=180
-        self.diagram = clg.linkage_print_diagram(linkage_swig_obj, 0, 180)
-        self.senses = clg.linkage_print_senses(linkage_swig_obj)
-        self.links_and_domains = clg.linkage_print_links_and_domains(linkage_swig_obj)
-        # Disable postscript. Lets be real, no one will use this.
-        # self.postscript_snippet = clg.linkage_print_postscript(linkage_swig_obj, 1, 0)
-        # self.postscript = clg.linkage_print_postscript(linkage_swig_obj, 1, 1)
+
+    def __init__(self, idx, sentence, parse_options):
+        self.sentence, self.parse_options = sentence, parse_options  # keep all args passed into clg.* fn
+        self._obj = clg.linkage_create(idx, sentence._obj, parse_options)
+
+    def __del__(self):
+        if hasattr(self, '_obj'):
+            clg.linkage_delete(self._obj)
+            del self._obj
+
+    def null_count(self):
+        return clg.sentence_null_count(self._obj)
+
+    def num_of_words(self):
+        return clg.linkage_get_num_words(self._obj)
+
+    def num_of_links(self):
+        return clg.linkage_get_num_links(self._obj)
+
+    def words(self):
+        for i in xrange(self.num_of_words()):
+            yield self.word(i)
+
+    def word(self, i):
+        return clg.linkage_get_word(self._obj, i)
+
+    def link_cost(self):
+        return clg.linkage_link_cost(self._obj)
+
+    def link(self, i):
+        return Link(self, i, self.word(clg.linkage_get_link_lword(self._obj, i)),
+                    clg.linkage_get_link_llabel(self._obj, i),
+                    clg.linkage_get_link_rlabel(self._obj, i),
+                    self.word(clg.linkage_get_link_rword(self._obj, i)))
+
+    def links(self):
+        for i in xrange(self.num_of_links()):
+            yield self.link(i)
+
+    def violation_name(self):
+        return clg.linkage_get_violation_name(self._obj)
+
+    def diagram(self, display_walls=0, screen_width=180):
+        return clg.linkage_print_diagram(self._obj, display_walls, screen_width)
+
+    def postscript(self, display_walls=1, print_ps_header=0):
+        return clg.linkage_print_postscript(self._obj, display_walls, print_ps_header)
+
+
+class Sentence(object):
+    text = None
+    dict = None
+    parse_options = None
+
+    def __init__(self, text, dict, parse_options):
+        """
+        In python 2.x, txt should be unicode string like u'Bo-bo-bo'
+        """
+        self.text, self.dict, self.parse_options = text, dict, parse_options  # keep all args passed into clg.* fn
+        self._obj = clg.sentence_create(self.text, self.dict._obj)
+
+    def __del__(self):
+        if hasattr(self, '_obj'):
+            clg.sentence_delete(self._obj)
+            del self._obj
+
+    def split(self):
+        return clg.sentence_split(self._obj)
+
+    def __len__(self):
+        return clg.sentence_length(self._obj)
```

Comment:
This is technically incorrect, and will need to change. The actual length of the sentence may vary, depending on the linkage.   So, for example: "We saw the seasand beach"  which is clearly mis-spelled .. but is it "we saw the seas and beach" (6 words) or is it "we saw the seasoned beach" (5 words)?   These are relatively rare, but the new tokenizer design allows these.

Similarly, the sentence no longer has a unique splitting into words, so the split() above cannot be used.